device/ArithmeticOperations.cpp

```diff
@@ -27,52 +40,115 @@ PackedFloat Multiply(PackedFloat const &a, PackedFloat const &b) {
     const ap_uint<kMantissaBits + 1> _m_mantissa =
         Karatsuba(a_mantissa_padded, b_mantissa_padded) >> (kMantissaBits - 1);
     // We need to shift the mantissa forward if the most significant bit is not set
-    const bool should_be_shifted = !IsLastBitSet(_m_mantissa);
+    const bool should_be_shifted = !IsMostSignificantBitSet(_m_mantissa);
     ap_uint<kMantissaBits + 1> m_mantissa = should_be_shifted ? _m_mantissa : (_m_mantissa >> 1);
     // Add up exponents. If the most significant bit was 1, we're done. Otherwise subtract 1 due to
     // the shift.
-    const Exponent m_exponent = a.exponent + b.exponent - should_be_shifted;
+    const Exponent m_exponent = a.exponent + b.exponent - (should_be_shifted ? 1 : 0);
 #endif
     // The sign is just the XOR of the existing signs
-    const bool m_sign = a.sign != b.sign;
-    return {m_sign, m_exponent, &m_mantissa};
+    PackedFloat result;
+    std::memcpy(result.mantissa, &m_mantissa, kMantissaBytes);
+    result.exponent = m_exponent;
+    result.sign = a.sign != b.sign;
+    return result;
 }
 
+// Does this correctly output the result if a and b are different signs?
+// The mantissa of the result should depend on the sign bits of a and b
```

Comment:
No, we only support positive, non-subnormal, non-special value numbers, since this should be all we need for the matmul.